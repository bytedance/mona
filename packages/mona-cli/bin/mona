#! /usr/bin/env node

const mona = require('../dist').default;
mona();
