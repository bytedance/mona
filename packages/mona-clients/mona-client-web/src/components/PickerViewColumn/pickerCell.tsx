// 参考arco Mobile
import React, { useCallback, useEffect, useMemo, useRef, CSSProperties, useState, useLayoutEffect } from 'react';
import { PickerData, ValueType, PickerCellMovingStatus } from '../Picker/type';
import styles from '../PickerView/index.module.less';
import { useRefState } from '../Picker/hooks';

export function getStyleWithVendor(style: any): CSSProperties {
  const allowReg = /(transform|transition|animation)/i;
  const newStyle = Object.keys(style).reduce<any>((acc, key) => {
    const webkitStyle = allowReg.test(key)
      ? {
          [`Webkit${key.replace(/^(.)/, (_, p1) => p1.toUpperCase())}`]: style[key],
        }
      : {};
    return {
      ...acc,
      [key]: style[key],
      ...webkitStyle,
    };
  }, {});
  return newStyle;
}

interface PickerCellProps {
  data: PickerData[];
  itemHeight: number;
  wrapperHeight: number;
  selectedValue?: ValueType;
  onValueChange?: (value: ValueType) => any;
  onScrollChange?: (value: ValueType) => any;
  rows?: number;
}

const PickerCell: React.FC<PickerCellProps> = props => {
  const { data, itemHeight, wrapperHeight, selectedValue, onValueChange, rows = 5, ...restProps } = props;

  const [transitionDuration, setTransitionDuration] = useState('');
  const [bezier, setBezier] = useState('');
  const [currentIndex, setCurrentIndex] = useState(0);
  const [currentValue, setCurrentValue] = useState(selectedValue);
  const [transformY, transformYRef, setTransformY] = useRefState(0);

  const lastTransformYRef = useRef(0);
  const touchStartTimeRef = useRef(0);
  const latestCallbackTimer = useRef(0);
  const touchStartYRef = useRef(0);
  const touchingRef = useRef(false);
  const wrapRef = useRef<HTMLDivElement | null>(null);

  const movingStatusRef = useRef<PickerCellMovingStatus>(PickerCellMovingStatus.normal);

  const rowCount = Math.max(rows % 2 === 0 ? rows + 1 : rows, 3);

  const colStyle = useMemo(
    () =>
      getStyleWithVendor({
        transform: `translate3d(0px, ${transformY || 0}px, 0px)`,
        ...(transitionDuration ? { transitionDuration } : {}),
        transitionTimingFunction: bezier,
        paddingBottom: `${((rowCount - 1) / 2) * itemHeight}px`,
        paddingTop: `${((rowCount - 1) / 2) * itemHeight}px`,
      }),
    [transitionDuration, transformY, bezier, itemHeight, rowCount],
  );

  function scrollingComplete(nowItemIndex: number) {
    if (currentIndex !== nowItemIndex) {
      setCurrentIndex(nowItemIndex);
      const newValue = data[nowItemIndex]?.value;

      if (newValue !== currentValue) {
        setCurrentValue(newValue);
        onValueChange?.(newValue);
      }
    }
  }

  const scrollTo = useCallback((transY: number, transDuration = 0, cb = () => {}) => {
    setTransitionDuration(transDuration ? `${transDuration}ms` : '');
    setTransformY(transY);

    if (latestCallbackTimer.current) {
      clearTimeout(latestCallbackTimer.current);
    }

    latestCallbackTimer.current = window.setTimeout(() => {
      movingStatusRef.current = PickerCellMovingStatus.normal;
      setTransitionDuration('');
      cb();
    }, transDuration);
  }, []);

  function scrollToIndex(itemIndex: number, transDuration = 0) {
    scrollTo(-1 * itemIndex * itemHeight, transDuration, () => {
      scrollingComplete(itemIndex);
    });
  }

  const handleColumnTouchStart = useCallback((e: TouchEvent) => {
    movingStatusRef.current = PickerCellMovingStatus.moving;
    const y = e.touches[0].screenY;
    touchStartTimeRef.current = Number(new Date());
    touchingRef.current = true;
    touchStartYRef.current = y;
    lastTransformYRef.current = transformYRef.current;
  }, []);

  const handleColumnTouchMove = useCallback(
    (e: TouchEvent) => {
      if (!touchingRef.current) {
        return;
      }
      e.cancelable && e.preventDefault();
      const lastTransformY = lastTransformYRef.current;
      const touchMoveY = e.touches[0].screenY;
      const distance = touchMoveY - touchStartYRef.current;
      const newPos = lastTransformY + distance;
      const maxPos = -1 * (data.length - 1) * itemHeight;
      setTransformY(
        (lastTransformY >= 0 && distance > 0) || (lastTransformY <= maxPos && distance < 0)
          ? lastTransformY + distance / 4
          : newPos,
      );
    },
    [data.length, itemHeight],
  );

  useEffect(() => {
    wrapRef.current?.addEventListener('touchstart', handleColumnTouchStart);
    wrapRef.current?.addEventListener('touchmove', handleColumnTouchMove);
    return () => {
      wrapRef.current?.removeEventListener('touchstart', handleColumnTouchStart);
      wrapRef.current?.removeEventListener('touchmove', handleColumnTouchMove);
    };
  }, [handleColumnTouchStart, handleColumnTouchMove]);

  function handleScrollEnd() {
    const maxIndex = data.length - 1;
    const nowIndex = Math.max(0, Math.min(maxIndex, Math.round((-1 * transformY) / itemHeight)));
    scrollToIndex(nowIndex, 100);
  }

  // 参考：https://juejin.im/post/6844904185121488910
  //@ts-ignore
  function momentum(current, start, duration, minY, maxY) {
    const durationMap: Record<string, number> = {
      noBounce: 800,
      weekBounce: 800,
      strongBounce: 400,
    };
    const bezierMap: Record<string, string> = {
      noBounce: 'cubic-bezier(.17, .89, .45, 1)',
      weekBounce: 'cubic-bezier(.25, .46, .45, .94)',
      strongBounce: 'cubic-bezier(.25, .46, .45, .94)',
    };
    let type = 'noBounce';
    const deceleration = 0.003;
    const bounceRate = 5;
    const bounceThreshold = 300;
    const maxOverflowY = wrapperHeight / 6;
    let overflowY;

    const distance = current - start;
    const speed = (2 * Math.abs(distance)) / duration;
    let destination = current + (speed / deceleration) * (distance < 0 ? -1 : 1);
    if (destination < minY) {
      overflowY = minY - destination;
      type = overflowY > bounceThreshold ? 'strongBounce' : 'weekBounce';
      destination = Math.max(minY - maxOverflowY, minY - overflowY / bounceRate);
    } else if (destination > maxY) {
      overflowY = destination - maxY;
      type = overflowY > bounceThreshold ? 'strongBounce' : 'weekBounce';
      destination = Math.min(maxY + maxOverflowY, maxY + overflowY / bounceRate);
    }

    return {
      destination,
      duration: durationMap[type],
      bezier: bezierMap[type],
    };
  }

  function handleColumnTouchEnd() {
    movingStatusRef.current = PickerCellMovingStatus.normal;
    const lastTransformY = lastTransformYRef.current;
    if (transformY === lastTransformY) {
      return;
    }
    touchingRef.current = false;
    const endTime = Number(new Date());
    const scrollerHeight = (data.length + rowCount - 1) * itemHeight;
    const duration = endTime - touchStartTimeRef.current;
    const absDistY = Math.abs(transformY - lastTransformY);
    if (duration < 300 && absDistY > 30) {
      const momentumY = momentum(transformY, lastTransformY, duration, wrapperHeight - scrollerHeight, 0);

      const newItemIndex = Math.max(
        0,
        Math.min(data.length - 1, Math.round((-1 * momentumY.destination) / itemHeight)),
      );

      setBezier(momentumY.bezier);

      movingStatusRef.current = PickerCellMovingStatus.scrolling;
      scrollToIndex(newItemIndex, momentumY.duration);
    } else {
      handleScrollEnd();
    }
  }

  function handleItemClick(itemIndex: number) {
    scrollToIndex(itemIndex, 100);
  }

  useLayoutEffect(() => {
    if ('selectedValue' in props) {
      const curIndex = data.findIndex((item: PickerData) => item.value === selectedValue);
      setCurrentIndex(curIndex);

      if (curIndex >= 0) {
        scrollToIndex(curIndex, 0);
      }
    }
    //@ts-ignore
  }, [selectedValue, itemHeight]);

  return data?.length ? (
    <div className={styles.pickerViewColumn} {...restProps}>
      <div
        className={styles.pickerViewColumnItemWrap}
        style={colStyle}
        ref={wrapRef}
        onTouchEnd={handleColumnTouchEnd}
        onTouchCancel={handleColumnTouchEnd}
      >
        {data.map((item, index) => {
          return (
            <div key={`${item.value}${index}`} style={{ height: itemHeight }} onClick={() => handleItemClick(index)}>
              {item.label}
            </div>
          );
        })}
      </div>
    </div>
  ) : null;
};

export default PickerCell;
