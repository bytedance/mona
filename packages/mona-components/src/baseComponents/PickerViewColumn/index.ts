import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps } from '../base';

export interface PickerViewColumnProps extends BaseProps {
  
}

export const PickerViewColumn: React.ComponentType<PickerViewColumnProps> = createComponent<PickerViewColumnProps>('picker-view-column')

