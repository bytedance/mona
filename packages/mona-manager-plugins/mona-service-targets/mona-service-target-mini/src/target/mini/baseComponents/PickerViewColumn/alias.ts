import { bubbleEventsAlias, basePropAlias } from '../prop';
import { PickerViewColumnAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
};

export default alias;
