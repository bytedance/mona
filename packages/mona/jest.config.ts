import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
