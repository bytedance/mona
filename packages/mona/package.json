{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
