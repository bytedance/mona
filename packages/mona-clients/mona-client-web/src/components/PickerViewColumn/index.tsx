import React from 'react';
import { PickerViewColumnProps } from '@bytedance/mona';

const PickerViewColumn: React.FC<PickerViewColumnProps> = ({ children }) => {
  return <div>{children}</div>
}

export default PickerViewColumn;
