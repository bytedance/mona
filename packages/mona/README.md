# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
