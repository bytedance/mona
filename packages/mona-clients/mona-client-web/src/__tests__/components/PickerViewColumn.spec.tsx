import React from 'react';
import { mount } from 'enzyme';
import PickerViewColumn from '../../components/PickerViewColumn';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { PickerViewColumnProps } from '@bytedance/mona';

describe('web component: PickerViewColumn', () => {
  mountTest(PickerViewColumn)
  handlerTest({ Component: PickerViewColumn, firstTag: 'div', props: { children: [<div key="1"></div>]} });

  it('should render correctly', () => {
    expect(mount(<PickerViewColumn />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: PickerViewColumnProps = {
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<PickerViewColumn {...props}>hello</PickerViewColumn>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
