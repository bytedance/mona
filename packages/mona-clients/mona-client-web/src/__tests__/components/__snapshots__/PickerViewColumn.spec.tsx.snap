// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: PickerViewColumn should have right properties 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "pickerViewColumn",
      "style": "color: red;",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "pickerViewColumnItemWrap",
          "style": "transform: translate3d(0px, 0px, 0px);",
        },
        "children": Array [
          Node {
            "attribs": Object {},
            "children": Array [
              Node {
                "data": "hello",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: PickerViewColumn should render correctly 1`] = `
LoadedCheerio {
  "0": Document {
    "children": Array [],
    "endIndex": null,
    "next": null,
    "parent": null,
    "prev": null,
    "startIndex": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
